### PR TITLE
Refactor the encoder/decoder and add more tests

### PR DESCRIPTION
**Pull request by asimonov-im**

### CODE REVIEW

**Comment by asimonov-im** [comment]
I tried to make the logic more readable - give it a look-see.

**Comment by MisterJimson** [comment]
I like the tests, its always great to have more.

Unfortunately I do not want to go forward with the code changes to the Decoder/Encoder. The reason being is that Valve's only spec for the Artifact Deck Code is the PHP code, no actual specification.

As they make changes and add more features or versions to the Artifact Deck Code, I would like to very easily implement those changes in C#. If our code structure differs from Valve's it will make it harder to stay in sync with their changes.

During my initial implementation I only made changes to the code where absolutely necessary for a C# port. If Valve released a spec with byte position information and stuff it would be different, but I don't want to derive a spec from PHP every time they change something.

**Comment by asimonov-im** [comment]
Apologies in advance for the long essay. I was originally just planning to close the PR with a simple "OK" but then decided to elaborate. 😸 

In my mind, every language has an ideal style and so an optimal C# implementation of an algorithm will not, and should not, look like its PHP counterpart. The goal of my refactor was to better understand how the encoding/decoding is done and to try to express those algorithms in more idiomatic C#.

Appearances can also be deceiving and just because two blocks of code look the same, doesn't mean they are. Take a look at how Valve implements truncating the deck name:
```
$nameLen = 0;
if( isset($deckContents['name']) )
{
	// replace strip_tags() with your own HTML santizer or escaper.
	$name = strip_tags( $deckContents['name'] );
	$trimLen = strlen($name);
	while( $trimLen > 63 )
	{
		$amountToTrim = floor( ($trimLen - 63) / 4 );
		$amountToTrim = ($amountToTrim > 1) ? $amountToTrim : 1;
		$name = mb_substr( $name, 0, mb_strlen($name) - $amountToTrim );
		$trimLen = strlen($name);
	}
	$nameLen = strlen($name);
}
```

Now look at the naive translation:
```
uint nameLen = 0;
string name = "";
if (deckContents.Name != null)
{
	// replace strip_tags() with your own HTML santizer or escaper.
	//name = strip_tags( deckContents['name']);
	name = deckContents.Name;
	int trimLen = name.Length;
	while (trimLen > 63)
	{
		int amountToTrim = (int)Math.Floor((decimal)(trimLen - 63) / 4);
		amountToTrim = (amountToTrim > 1) ? amountToTrim : 1;
		name = name.Substring(0, name.Length - amountToTrim);
		trimLen = name.Length;
	}

	nameLen = (uint)name.Length;
}
```

This is functionally equivalent to:
```
string name = deckContents.Name ?? string.Empty;
name = name.Substring(0, Math.Min(63, name.Length));
```

It look me about an hour of scratching my head to figure out why Valve's algorithm to truncate a string looks so ugly and complicated. It's because they are trimming it to be representable in **63 bytes in UTF8**, which requires handling multibyte characters correctly. The C# implementation doesn't work the same way because String.Length has different behavior than mb_substr() or substr() in PHP. On a related note, repeatedly calling String.Substring() is also not very efficient.

For example:
```
string koreanName = "이것은 전 세계와 아마도 우주에서 최고의 카드 덱입니다";
int length = koreanName.Length; // 30
int utf8ByteLenth = Encoding.UTF8.GetByteCount(koreanName); // 76
```

In summary, I accept your reasoning although I do feel a bit disappointed that my efforts were wasted (they weren't actually, since I'll just maintain them on my fork). I **would** caution you against trying to stay too close to Valve's code because it might cause you to miss out on future improvements and discourage future contributions from other developers.

**Comment by MisterJimson** [comment]
I 100% agree with you, the only reason I am keeping it the same as PHP is to merge changes made to Valve's repo over time.

We can make the Encoder/Decoder a lot better in C#, but as soon as Valve makes a change to the PHP code it will be very difficult to merge those changes in.

If there are any changes to the Encoder/Decoder that preserve the code flow, structure, and naming I will gladly accept them.

**Comment by asimonov-im** [comment]
To allow myself the freedom to stray from Valve's implementation, I have diverged my fork of the project and renamed it to [ArtifactDeckSharp](https://github.com/asimonov-im/ArtifactDeckSharp).

Finally, I'd like to thank you for all of the work you've done so far. Without it, I definitely would not have gotten so interested in the project.

**Comment by MisterJimson** [comment]
Cool, and no worries man. Hopefully we can share some cool features over time.
